### PR TITLE
CDPCP-1212 Update password expiration for hashed password codepath

### PR DESCRIPTION
**Pull request by handavid**
The UMS now provides a password expiration time along with the
workload credentials. This value is epoch millis. User sync has
been updated to set this expiration time when setting the hashed
password. The userMod request was refactored, moving the param
map creation into the FreeIpaClient to match the pattern used by
other FreeIpaClient methods.

The UsersState was slightly refactored to split out the UMS-specific
information into a UmsUsersState because the UsersState model is
also used to represent the ipa state.

The Mock user management service has been modified to return an
expiration time that is different from the max expiration time
that the FMS uses.

### CODE REVIEW

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -3,8 +3,10 @@
 
 <module name="Checker">
     <property name="severity" value="error" />
+    <module name="SuppressWarningsFilter" />
     <module name="TreeWalker">
         <property name="tabWidth" value="4" />
+        <module name="SuppressWarningsHolder" />
```
Comment (giladwolff) [inline config/checkstyle/checkstyle.xml]: For my edification, why these changes?

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java [hunk 2]
```diff
@@ -150,7 +156,7 @@ public User userAdd(String user, String firstName, String lastName) throws FreeI
                 "sn", lastName,
                 "loginshell", "/bin/bash",
                 "random", true,
-                "setattr", "krbPasswordExpiration=" + PASSWORD_EXPIRATION_DATETIME
+                "setattr", "krbPasswordExpiration=" + MAX_PASSWORD_EXPIRATION_DATETIME
```
Comment (giladwolff) [inline freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: again, for my edification, why are we setting the password expiration when we add the user?

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java [hunk 3]
```diff
@@ -159,11 +165,34 @@ public void userSetPassword(String user, String password) throws FreeIpaClientEx
         // FreeIPA expires any password that is set by another user. Work around this by
         // performing a separate API call to set the password expiration into the future
         userMod(user, "userpassword", password);
-        updateUserPasswordExpiration(user);
+        updateUserPasswordMaxExpiration(user);
+    }
+
+    public void updateUserPasswordExpiration(String user, long expiration) throws FreeIpaClientException {
+        String passwordExpirationDate = formatDate(expiration);
+        userMod(user, "setattr", "krbPasswordExpiration=" + passwordExpirationDate);
```
Comment (giladwolff) [inline freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: So above you set the krbPasswordExpiration with a date time string (with a certain pattern). Here it's just a long. Are both correct? If they are, let's settle on one way to identity the expiration time. Personally I like to use java Instant and Duration to hold and calculate time - it is more robust than passing along longs.

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java [hunk 4]
```diff
@@ -159,11 +165,34 @@ public void userSetPassword(String user, String password) throws FreeIpaClientEx
...
+    }
+
+    public void updateUserPasswordMaxExpiration(String user) throws FreeIpaClientException {
+        updateUserPasswordExpiration(user, 0);
```
Comment (giladwolff) [inline freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: same question. Here you pass 0, does that mean that the password never expires? Also, maybe have a readability constant for this?

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java [hunk 5]
```diff
@@ -159,11 +165,34 @@ public void userSetPassword(String user, String password) throws FreeIpaClientEx
...
+        Map<String, Object> params =
+                Map.of("setattr", List.of(
+                        "cdpHashedPassword=" + hashedPassword,
+                        "cdpUnencryptedKrbPrincipalKey=" + unencryptedKrbPrincipalKey,
+                        "krbPasswordExpiration=" + passwordExpirationDate));
+        return userMod(user, params);
+    }
+
+    String formatDate(long timestampMillis) {
+        if (timestampMillis <= 0) {
```
Comment (giladwolff) [inline freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: I think it's safer if you assert that timestampMillis < 0. It's a bug.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaPostInstallService.java [hunk 6]
```diff
@@ -80,11 +80,11 @@ public void postInstallFreeIpa(Long stackId) throws Exception {
 
     private void modifyAdminPasswordExpirationIfNeeded(FreeIpaClient client) throws FreeIpaClientException {
         Optional<User> user = client.userFind(freeIpaClientFactory.getAdminUser());
-        if (user.isPresent() && !FreeIpaClient.PASSWORD_EXPIRATION_DATETIME.equals(user.get().getKrbPasswordExpiration())) {
+        if (user.isPresent() && !FreeIpaClient.MAX_PASSWORD_EXPIRATION_DATETIME.equals(user.get().getKrbPasswordExpiration())) {
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaPostInstallService.java]: So the admin password is our own freeipa admin user? and this is to ensure that the password never expires?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java [hunk 7]
```diff
@@ -120,7 +122,8 @@ private WorkloadCredential getCredentials(String userCrn, Optional<String> reque
         GetActorWorkloadCredentialsResponse response = grpcUmsClient.getActorWorkloadCredentials(IAM_INTERNAL_ACTOR_CRN, userCrn, requestId);
         String hashedPassword = response.getPasswordHash();
         List<ActorKerberosKey> keys = response.getKerberosKeysList();
-        return new WorkloadCredential(hashedPassword, keys);
+        long expirationDate = response.getPasswordHashExpirationDate();
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java]: Suggestion: convert it to an Instant here.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
+package com.sequenceiq.freeipa.service.freeipa.user.model;
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java]: Feel free to ignore if this is not the coding convention in this repo, but copywrong statement is missing.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class UmsUsersState {
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java]: javadoc?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+    private final UsersState usersState;
+
+    private final Map<String, WorkloadCredential> usersWorkloadCredentialMap;
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java]: Can this be immutable?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    private final Set<FmsUser> requestedWorkloadUsers;
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java]: Can this be immutable?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+    public UmsUsersState(UsersState usersState, Map<String, WorkloadCredential> usersWorkloadCredentialMap, Set<FmsUser> requestedWorkloadUsers) {
+        this.usersState = requireNonNull(usersState, "UsersState is null");
+        this.usersWorkloadCredentialMap = requireNonNull(usersWorkloadCredentialMap, "workload credential map is null");
+        this.requestedWorkloadUsers = requireNonNull(requestedWorkloadUsers, "requested workload users is null");
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java]: I would make a copy of the map and set, even if you don't make the members immutable.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+            this.usersState = usersState;
+        }
+
+        public void addWorkloadCredentials(String userName, WorkloadCredential creds) {
+            workloadCredentialMap.put(userName, creds);
+        }
+
+        public void addRequestedWorkloadUsers(FmsUser user) {
+            requestedWorkloadUsers.add(user);
+        }
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java]: builders' methods usually returns with `this` so you can chain methods

File: config/checkstyle/checkstyle.xml (hunk 1 above)
Comment (handavid) [reply config/checkstyle/checkstyle.xml]: these changes allow me to use the @SuppressWarnings annotation on one method that has 8 parameters. `UmsUsersStateProvider.handleUser()`

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java (hunk 2 above)
Comment (handavid) [reply freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: We set a random password so that we can generate tgts. I think that we had to set this so that the random password isn't expired.

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java (hunk 3 above)
Comment (handavid) [reply freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: all expirations are currently passed through the method parameters as longs. They are converted to the map {"setattr":"krbPasswordExpiration=1234567890Z"} when constructing the ipa JSON-RPC request. The userMod() method doesn't know anything about password expiration.

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java (hunk 4 above)
Comment (handavid) [reply freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: I'll add a constant

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java (hunk 5 above)
Comment (handavid) [reply freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: will do

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaPostInstallService.java (hunk 6 above)
Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaPostInstallService.java]: yes. otherwise the FMS won't be able to sync users or set passwords after some time in the future.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java (hunk 7 above)
Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java]: converted it to an Optional<Instant> so we can do the 0-checking here where the 0 is meaningful.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java (hunk 10 above)
Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java]: done

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java (hunk 11 above)
Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java]: done

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java (hunk 13 above)
Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UmsUsersState.java]: done

**Comment by handavid** [comment]
> super nit: you can declare them as ImmutableMap/ImmutableSet.

I'll add that to my next PR
